### PR TITLE
Create histograms approximations with adhoc bins.

### PR DESCRIPTION
**Pull request by postwait**

### CODE REVIEW

File: src/circllhist.c [hunk 1]
```diff
@@ -1637,3 +1641,42 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+    switch(mode) {
+      case HIST_APPROX_HIGH:
+        m = bins[i].upper - fabs(bins[i].upper) / 1000; break;
+      case HIST_APPROX_LOW:
+        m = bins[i].lower + fabs(bins[i].lower) / 1000; break;
+      case HIST_APPROX_MID:
+        m = (a / (a + b)) * fabs(b-a) + bins[i].lower;
+        break;
+    }
+    // however if we're a (-inf,B) or an (A,+inf) bucket we need to me more clever
```
Comment (redhotpenguin) [inline src/circllhist.c:1669]: s/me/be/

Custom bins layout?

Comment (postwait) [reply src/circllhist.c]: Thanks.  Support for consuming the "oddly binned"

File: src/circllhist.c [hunk 2]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+    }
+    if(i == 0 && nbins > 2 && bins[0].lower <= -pow(10,128)) {
+      double offset1 = abs(b), offset2 = abs(bins[1].upper - bins[1].lower);
+      m = b - (offset1 > offset2 ? offset2 : offset1);
+    }
+    hist_insert(h, m, bins[i].count);
+    last_spread = b - a;
+  }
+  return h;
+}
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1685]: Tests would be nice : ) 

The circllhist has 5 relevant regions:

A = [-inf, -e+128) -- Min bin
B = (-e+128, -e-128] -- Negative significant range
C = (-e-128, +e-128) -- Zero bin
D = (+e-128, +e+128] -- Positive significant range
E = [e+128, inf) -- Min bin

The sides lower,upper of the adhoc bin could lay anywhere in those 5 areas. 

I am not sure which combinations are intended to be supported by this method, nor how the method would behave in all those cases.

File: src/circllhist.c [hunk 3]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+                                     size_t nbins, double sum) {
+  histogram_t *h = hist_alloc_nbins(nbins);
+  double last_spread = 0;
+  for(size_t i=0; i<nbins; i++) {
+    double a = bins[i].lower;
+    double b = bins[i].upper;
+    if(abs(a) > abs(b)) {
+      a = bins[i].upper;
+      b = bins[i].lower;
+    }
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1653]: What's the intention behind this renaming? Is that to take care of the negative range? Does this work when one of the sides is negative and not the other (e.g. a = -1, b=0.01) ?

File: src/circllhist.c [hunk 4]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+    // chose a midpoint with minimal error
+    double m = 0;
+    switch(mode) {
+      case HIST_APPROX_HIGH:
+        m = bins[i].upper - fabs(bins[i].upper) / 1000;
+        break;
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1659]: This seems to work, but left me confused for a while (since m does not have to be inside the ad-hoc bin).
More natural would have been to use a "convex combination" like so:

```m = 0.001 * bins[i].lower + 0.999 * bins[i].upper // select a point the upper end of the ad-hoc bin.```

File: src/circllhist.c [hunk 5]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+      case HIST_APPROX_LOW:
+        m = bins[i].lower + fabs(bins[i].lower) / 1000;
+        break;
+      case HIST_APPROX_MID:
+        m = a + (b-a)/2;
+        break;
+      case HIST_APPROX_MIN_ERROR:
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1666]: As Neil Gunther pointed out this is the "harmonic mean". We could adopt this naming.

File: src/circllhist.c [hunk 6]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
   }
   return 0;
 }
+
+histogram_t *
+hist_create_approximation_from_adhoc(histogram_approx_mode_t mode,
+                                     const histogram_adhoc_bin_t *bins,
+                                     size_t nbins, double sum) {
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1644]: `sum` is unused.

File: src/circllhist.c [hunk 7]
```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
...
+        break;
+    }
+    // however if we're a (-inf,B) or an (A,+inf) bucket we need to be more clever
+    if(i>0 && bins[i].upper >= pow(10,128)) {
+      m = a + (last_spread > a ? a : last_spread);
+    }
+    if(i == 0 && nbins > 2 && bins[0].lower <= -pow(10,128)) {
+      double offset1 = abs(b), offset2 = abs(bins[1].upper - bins[1].lower);
+      m = b - (offset1 > offset2 ? offset2 : offset1);
+    }
```
Comment (HeinrichHartmann) [inline src/circllhist.c:1680]: Not sure about those special cases. (I don't see a problem, but I it's not clear to me what this is doing).

File: src/circllhist.c (hunk 4 above)
Comment (postwait) [reply src/circllhist.c:1659]: I like this.

File: src/circllhist.c (hunk 7 above)
Comment (postwait) [reply src/circllhist.c:1680]: This is the handle the case where one end of the interval is 0 or 0-like.  The harmonic mean in that case is not benefiting the user... 0.

**Comment by HeinrichHartmann** [comment]
Looks like the tests did catch a few more edge cases. LGTM!
